File: frontend/js/portfolio-advisor.js
/**
 * Avatar Base — Portfolio Advisor Engine (Frontend)
 * SEBI-compliant drift detection, concentration alerts, suitability gate
 */

const DRIFT_THRESHOLD = 5.0;           // >5% triggers alert
const CONCENTRATION_THRESHOLD = 20.0; // >20% single stock/sector triggers alert

const SEBI_DISCLOSURE = `⚠️ REGULATORY DISCLOSURE: Investments in securities market are subject to market risks. Read all scheme-related documents carefully before investing. Past performance is not indicative of future returns. Avatar Base is an AI system — not a SEBI-registered Investment Adviser (IA). For personalised advice, consult a SEBI-registered IA. Stock recommendations originate exclusively from SEBI-registered Research Analysts.`;

const RISK_ORDER = { CONSERVATIVE: 1, MODERATE: 2, AGGRESSIVE: 3, GROWTH: 4, LOW: 1, MEDIUM: 2, HIGH: 3, VERY_HIGH: 4 };

// ── Asset Allocation Computation ──────────────────────────────────────
export function computeAssetAllocation(holdings) {
  const totals = {};
  let grandTotal = 0;
  for (const h of holdings) {
    const ac = h.assetClass || 'UNKNOWN';
    totals[ac] = (totals[ac] || 0) + (h.currentValue || 0);
    grandTotal += (h.currentValue || 0);
  }
  if (grandTotal === 0) return {};
  return Object.fromEntries(
    Object.entries(totals).map(([k, v]) => [k, parseFloat((v / grandTotal * 100).toFixed(2))])
  );
}

// ── Drift Alert Detection ─────────────────────────────────────────────
export function detectDriftAlerts(currentAlloc, targetAlloc) {
  const alerts = [];
  for (const [assetClass, target] of Object.entries(targetAlloc)) {
    const current = currentAlloc[assetClass] || 0;
    const drift = current - target;
    if (Math.abs(drift) > DRIFT_THRESHOLD) {
      const direction = drift > 0 ? 'OVERWEIGHT' : 'UNDERWEIGHT';
      alerts.push({
        assetClass, current, target, drift: parseFloat(drift.toFixed(2)),
        direction,
        severity: Math.abs(drift) > 10 ? 'HIGH' : 'MEDIUM',
        action: direction === 'OVERWEIGHT'
          ? `Trim ${assetClass} by ~${Math.abs(drift).toFixed(1)}% — consider profit-booking into underweight classes.`
          : `Top up ${assetClass} by ~${Math.abs(drift).toFixed(1)}% to restore target allocation.`,
        sebiDisclosure: SEBI_DISCLOSURE,
      });
    }
  }
  return alerts;
}

// ── Concentration Risk Detection ──────────────────────────────────────
export function detectConcentrationRisks(holdings) {
  const totalValue = holdings.reduce((s, h) => s + (h.currentValue || 0), 0);
  if (totalValue === 0) return [];

  const alerts = [];
  const stockMap = {};
  const sectorMap = {};

  for (const h of holdings) {
    if (h.instrumentType !== 'EQUITY') continue;
    const pct = h.currentValue / totalValue * 100;
    const isin = h.isin || h.name;
    stockMap[isin] = { name: h.name, pct: (stockMap[isin]?.pct || 0) + pct };
    const sector = h.sector || 'Unknown';
    sectorMap[sector] = (sectorMap[sector] || 0) + pct;
  }

  for (const [isin, info] of Object.entries(stockMap)) {
    if (info.pct > CONCENTRATION_THRESHOLD) {
      alerts.push({
        type: 'SINGLE_STOCK', identifier: isin, name: info.name,
        pct: parseFloat(info.pct.toFixed(2)), threshold: CONCENTRATION_THRESHOLD,
        severity: info.pct > 30 ? 'HIGH' : 'MEDIUM',
        action: `${info.name} is ${info.pct.toFixed(1)}% of portfolio (safe limit: ${CONCENTRATION_THRESHOLD}%). Consider partial profit-booking.`,
        sebiDisclosure: SEBI_DISCLOSURE,
      });
    }
  }

  for (const [sector, pct] of Object.entries(sectorMap)) {
    if (pct > CONCENTRATION_THRESHOLD) {
      alerts.push({
        type: 'SECTOR', identifier: sector, name: sector,
        pct: parseFloat(pct.toFixed(2)), threshold: CONCENTRATION_THRESHOLD,
        severity: pct > 35 ? 'HIGH' : 'MEDIUM',
        action: `${sector} exposure (${pct.toFixed(1)}%) exceeds ${CONCENTRATION_THRESHOLD}% sector limit. Diversify across sectors.`,
        sebiDisclosure: SEBI_DISCLOSURE,
      });
    }
  }
  return alerts;
}

// ── Execution Suitability Gate ────────────────────────────────────────
export function checkSuitability({ userRiskProfile, instrumentRiskLevel, sebiRaValidated }) {
  // Hard compliance block if not RA-validated
  if (!sebiRaValidated) {
    return {
      isSuitable: false,
      canOverride: false,
      blockType: 'COMPLIANCE_HARD_BLOCK',
      message: '🚫 COMPLIANCE BLOCK: This instrument has not been validated by a SEBI-registered Research Analyst. Transaction cannot proceed. Avatar Base only recommends RA-approved instruments.',
      sebiDisclosure: SEBI_DISCLOSURE,
    };
  }

  const userTolerance = RISK_ORDER[userRiskProfile] || 0;
  const instrumentRisk = RISK_ORDER[instrumentRiskLevel] || 0;

  if (instrumentRisk > userTolerance) {
    return {
      isSuitable: false,
      canOverride: true,
      blockType: 'SUITABILITY_MISMATCH',
      message: `⚠️ Suitability mismatch: This instrument is rated '${instrumentRiskLevel}' risk but your profile is '${userRiskProfile}'. You may proceed with explicit acknowledgment of the additional risk.`,
      sebiDisclosure: SEBI_DISCLOSURE,
    };
  }

  return { isSuitable: true, canOverride: false, blockType: null, message: null, sebiDisclosure: SEBI_DISCLOSURE };
}

// ── Portfolio Rebalancing Suggestions ─────────────────────────────────
export function generateRebalancingSuggestions(driftAlerts, holdings) {
  return driftAlerts.map(alert => {
    const targetChange = Math.abs(alert.drift);
    return {
      assetClass: alert.assetClass,
      direction: alert.direction,
      changeRequired: `${targetChange.toFixed(1)}%`,
      suggestedAction: alert.action,
      sebiDisclosure: SEBI_DISCLOSURE,
      requiresAuth: 2,
    };
  });
}

export { SEBI_DISCLOSURE };

File: backend/services/portfolio_advisor.py
from typing import List, Dict, Tuple, Optional

SEBI_DISCLOSURE = (
    "⚠️ REGULATORY DISCLOSURE (SEBI): Investments in securities market are subject to market risks. "
    "Read all scheme-related documents carefully before investing. Past performance is not indicative of future returns. "
    "This is an AI-generated advisory suggestion from Avatar Base. Avatar Base is not a SEBI-registered Investment Adviser. "
    "For personalised advice, consult a SEBI-registered Investment Adviser. "
    "SEBI Registration: [IA-XXXXX] — Research Analyst: [RA-XXXXXX]."
)

DRIFT_THRESHOLD_PCT = 5.0
CONCENTRATION_THRESHOLD_PCT = 20.0

# Ordered order of risk profiles
RISK_ORDER = {
    "CONSERVATIVE": 1,
    "MODERATE": 2,
    "AGGRESSIVE": 3,
    "GROWTH": 4,
    "LOW": 1,
    "MEDIUM": 2,
    "HIGH": 3,
    "VERY_HIGH": 4
}

def detect_drift(current_allocation: Dict[str, float], target_allocation: Dict[str, float], threshold: float = DRIFT_THRESHOLD_PCT) -> List[Dict]:
    """
    Checks if any asset class allocation deviates from the target by more than the threshold.
    """
    alerts = []
    for asset_class, target_pct in target_allocation.items():
        curr_pct = current_allocation.get(asset_class, 0.0)
        drift = curr_pct - target_pct
        if abs(drift) >= threshold:
            direction = "OVERWEIGHT" if drift > 0 else "UNDERWEIGHT"
            action = (
                f"Trim {asset_class} by ~{abs(drift):.1f}% and reallocate profits into underweight assets."
                if direction == "OVERWEIGHT"
                else f"Top up {asset_class} by ~{abs(drift):.1f}% to align with target allocations."
            )
            alerts.append({
                "asset_class": asset_class,
                "current_pct": curr_pct,
                "target_pct": target_pct,
                "drift_pct": round(drift, 2),
                "direction": direction,
                "action": action,
                "sebi_disclosure": SEBI_DISCLOSURE
            })
    return alerts

def detect_concentration_risks(holdings: List[Dict], total_portfolio_value: float) -> List[Dict]:
    """
    Checks if single stock or sector concentration exceeds the 20% limit.
    """
    if total_portfolio_value == 0:
        return []
        
    alerts = []
    stock_values: Dict[str, Dict] = {}
    sector_values: Dict[str, float] = {}
    
    for h in holdings:
        val = h.get("current_value_paise", 0.0) / 100.0 # Convert to Rs
        pct = (val / total_portfolio_value) * 100.0
        
        # Track single stocks
        if h.get("instrument_type") == "EQUITY":
            isin = h.get("isin", h.get("instrument_name", "UNKNOWN"))
            if isin not in stock_values:
                stock_values[isin] = {"name": h.get("instrument_name"), "pct": 0.0}
            stock_values[isin]["pct"] += pct
            
            # Track sectors
            sector = h.get("sector", "UNKNOWN")
            sector_values[sector] = sector_values.get(sector, 0.0) + pct
            
    # Stock Concentration Alert
    for isin, info in stock_values.items():
        if info["pct"] > CONCENTRATION_THRESHOLD_PCT:
            alerts.append({
                "type": "SINGLE_STOCK",
                "identifier": isin,
                "display_name": info["name"],
                "current_pct": round(info["pct"], 2),
                "action": f"Single stock '{info['name']}' is {info['pct']:.1f}% of your portfolio. Consider booking profits to diversify."
            })
            
    # Sector Concentration Alert
    for sector, pct in sector_values.items():
        if pct > CONCENTRATION_THRESHOLD_PCT:
            alerts.append({
                "type": "SECTOR",
                "identifier": sector,
                "display_name": sector,
                "current_pct": round(pct, 2),
                "action": f"Sector '{sector}' is {pct:.1f}% of your portfolio. Diversify across other sectors to reduce systemic risk."
            })
            
    return alerts

def check_execution_suitability(
    user_risk_profile: str,
    instrument_risk_level: str,
    instrument_isin: str,
    sebi_ra_validated: bool
) -> Dict:
    """
    Blocks transactions if equity picks are not SEBI-RA validated.
    Warns users if there is a suitability mismatch.
    """
    if not sebi_ra_validated:
        return {
            "is_suitable": False,
            "can_override": False,
            "block_reason": (
                "COMPLIANCE BLOCK: This instrument has not been validated by a SEBI-registered Research Analyst. "
                "Advisory execution blocked."
            ),
            "sebi_disclosure": SEBI_DISCLOSURE
        }
        
    user_tol = RISK_ORDER.get(user_risk_profile.upper(), 1)
    inst_risk = RISK_ORDER.get(instrument_risk_level.upper(), 1)
    
    if inst_risk > user_tol:
        return {
            "is_suitable": False,
            "can_override": True,
            "block_reason": (
                f"Suitability warning: This instrument is rated '{instrument_risk_level}' risk, "
                f"but your risk profile is '{user_risk_profile}'. Confirm to override with risk acknowledgment."
            ),
            "sebi_disclosure": SEBI_DISCLOSURE
        }
        
    return {
        "is_suitable": True,
        "can_override": False,
        "block_reason": None,
        "sebi_disclosure": SEBI_DISCLOSURE
    }
